# EXERCICI 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [1]:
import pandas as pd
airlines_delay = pd.read_csv('/Users/angel/Desktop/DelayedFlights.csv')
airlines_delay

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

In [2]:
print(airlines_delay[airlines_delay.duplicated()])

Empty DataFrame
Columns: [Unnamed: 0, Year, Month, DayofMonth, DayOfWeek, DepTime, CRSDepTime, ArrTime, CRSArrTime, UniqueCarrier, FlightNum, TailNum, ActualElapsedTime, CRSElapsedTime, AirTime, ArrDelay, DepDelay, Origin, Dest, Distance, TaxiIn, TaxiOut, Cancelled, CancellationCode, Diverted, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay]
Index: []

[0 rows x 30 columns]


Compruebo que no hay duplicados

### VALORES NaN

In [3]:
print (airlines_delay.isnull().sum())

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64


Selecciono las filas con valores nulos en AirTime (tiempo que el avión pasa en el aire) y ArrDelay (retraso sobre el tiempo previsto de llegada del avión).

In [4]:
filas_nan = airlines_delay[airlines_delay['AirTime'].isnull() & airlines_delay['ArrDelay'].isnull()]
filas_nan

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1280           1763  2008      1           3          4    922.0         915   
1372           1911  2008      1           3          4   2325.0        1900   
1776           2651  2008      1           4          5   1949.0        1905   
1831           2726  2008      1           4          5    737.0         705   
2244           3672  2008      1           4          5   1849.0        1630   
...             ...   ...    ...         ...        ...      ...         ...   
1935651     7006289  2008     12          10          3   1459.0        1447   
1935716     7006401  2008     12          11          4   1355.0        1106   
1935876     7006809  2008     12          11          4   1026.0         955   
1935978     7007034  2008     12          11          4   1527.0        1520   
1936470     7008584  2008     12          12          5    703.0         630   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1280         NaN        1050            WN  ...     NaN    12.0          0   
1372         NaN        2030            WN  ...     NaN    11.0          0   
1776         NaN        1910            WN  ...     NaN    11.0          0   
1831         NaN         825            WN  ...     NaN    13.0          0   
2244         NaN        1755            WN  ...     NaN    12.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1935651      NaN        1650            DL  ...     NaN    37.0          1   
1935716      7.0        1950            DL  ...    13.0    17.0          0   
1935876      NaN        1219            DL  ...     NaN     NaN          1   
1935978   2106.0        1708            DL  ...     9.0    19.0          0   
1936470      NaN         734            DL  ...     NaN    33.0          1   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
1280                    N         1           NaN           NaN      NaN   
1372                    N         1           NaN           NaN      NaN   
1776                    N         1           NaN           NaN      NaN   
1831                    N         1           NaN           NaN      NaN   
2244                    N         1           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1935651                 A         0           NaN           NaN      NaN   
1935716                 N         1           NaN           NaN      NaN   
1935876                 A         0           NaN           NaN      NaN   
1935978                 N         1           NaN           NaN      NaN   
1936470                 B         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
1280              NaN                NaN  
1372              NaN                NaN  
1776              NaN                NaN  
1831              NaN                NaN  
2244              NaN                NaN  
...               ...                ...  
1935651           NaN                NaN  
1935716           NaN                NaN  
1935876           NaN                NaN  
1935978           NaN                NaN  
1936470           NaN                NaN  

[8387 rows x 30 columns]

In [5]:
cancelados_o_desviados = filas_nan[(filas_nan['Cancelled'] == 1) | (filas_nan['Diverted'] == 1)]
cancelados_o_desviados

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1280           1763  2008      1           3          4    922.0         915   
1372           1911  2008      1           3          4   2325.0        1900   
1776           2651  2008      1           4          5   1949.0        1905   
1831           2726  2008      1           4          5    737.0         705   
2244           3672  2008      1           4          5   1849.0        1630   
...             ...   ...    ...         ...        ...      ...         ...   
1935651     7006289  2008     12          10          3   1459.0        1447   
1935716     7006401  2008     12          11          4   1355.0        1106   
1935876     7006809  2008     12          11          4   1026.0         955   
1935978     7007034  2008     12          11          4   1527.0        1520   
1936470     7008584  2008     12          12          5    703.0         630   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1280         NaN        1050            WN  ...     NaN    12.0          0   
1372         NaN        2030            WN  ...     NaN    11.0          0   
1776         NaN        1910            WN  ...     NaN    11.0          0   
1831         NaN         825            WN  ...     NaN    13.0          0   
2244         NaN        1755            WN  ...     NaN    12.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1935651      NaN        1650            DL  ...     NaN    37.0          1   
1935716      7.0        1950            DL  ...    13.0    17.0          0   
1935876      NaN        1219            DL  ...     NaN     NaN          1   
1935978   2106.0        1708            DL  ...     9.0    19.0          0   
1936470      NaN         734            DL  ...     NaN    33.0          1   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
1280                    N         1           NaN           NaN      NaN   
1372                    N         1           NaN           NaN      NaN   
1776                    N         1           NaN           NaN      NaN   
1831                    N         1           NaN           NaN      NaN   
2244                    N         1           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1935651                 A         0           NaN           NaN      NaN   
1935716                 N         1           NaN           NaN      NaN   
1935876                 A         0           NaN           NaN      NaN   
1935978                 N         1           NaN           NaN      NaN   
1936470                 B         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
1280              NaN                NaN  
1372              NaN                NaN  
1776              NaN                NaN  
1831              NaN                NaN  
2244              NaN                NaN  
...               ...                ...  
1935651           NaN                NaN  
1935716           NaN                NaN  
1935876           NaN                NaN  
1935978           NaN                NaN  
1936470           NaN                NaN  

[8387 rows x 30 columns]

Coinciden los vuelos cancelados o desviados con los que no tienen datos en AirTime y en ArrDelay. Elimino estas filas porque al ser vuelos cancelados o desviados pueden producir valores atípicos en los cálculos.

In [6]:
airlines_delay_no_nan = airlines_delay.drop(filas_nan.index)
airlines_delay_no_nan

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1928371 rows x 30 columns]

Con esta operación se eliminan 8387 entradas, representando el 0.43% del total de la muestra y sabemos que los vuelos que quedan no han sido cancelados o desviados.

In [7]:
print (airlines_delay_no_nan.isnull().sum())

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                   0
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   3
ActualElapsedTime         0
CRSElapsedTime            0
AirTime                   0
ArrDelay                  0
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                    0
TaxiOut                   0
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         680883
WeatherDelay         680883
NASDelay             680883
SecurityDelay        680883
LateAircraftDelay    680883
dtype: int64


Al eliminar cancelados y desviados desaparecen NaN en casi todas las categorías excepto las que clasifican de quién es consecuencia el retraso. 
También hay un valor de 3 NaN en TailNum (identificador único del avión). 
Para eliminar todos los datos faltantes hay que suprimir 680883 entradas lo que representa el 35.16% del total del conjunto. Es una cantidad considerable, pero la falta de información de las causas del retraso podría llevar a un análisis incompleto.

In [8]:
airlines_delay_no_nan = airlines_delay_no_nan.dropna()
airlines_delay_no_nan

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
3                 4  2008      1           3          4   1829.0        1755   
5                 6  2008      1           3          4   1937.0        1830   
7                11  2008      1           3          4   1644.0        1510   
9                16  2008      1           3          4   1452.0        1425   
11               18  2008      1           3          4   1323.0        1255   
...             ...   ...    ...         ...        ...      ...         ...   
1936751     7009705  2008     12          13          6    921.0         830   
1936752     7009709  2008     12          13          6   1552.0        1520   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
3         1959.0        1925            WN  ...     3.0    10.0          0   
5         2037.0        1940            WN  ...     3.0     7.0          0   
7         1845.0        1725            WN  ...     6.0     8.0          0   
9         1640.0        1625            WN  ...     7.0     8.0          0   
11        1526.0        1510            WN  ...     4.0     9.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936751   1112.0        1008            DL  ...     8.0    21.0          0   
1936752   1735.0        1718            DL  ...     9.0     7.0          0   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
3                       N         0           2.0           0.0      0.0   
5                       N         0          10.0           0.0      0.0   
7                       N         0           8.0           0.0      0.0   
9                       N         0           3.0           0.0      0.0   
11                      N         0           0.0           0.0      0.0   
...                   ...       ...           ...           ...      ...   
1936751                 N         0          51.0           0.0     13.0   
1936752                 N         0           0.0           0.0      0.0   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   

        SecurityDelay  LateAircraftDelay  
3                 0.0               32.0  
5                 0.0               47.0  
7                 0.0               72.0  
9                 0.0               12.0  
11                0.0               16.0  
...               ...                ...  
1936751           0.0                0.0  
1936752           0.0               17.0  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  

[1247486 rows x 30 columns]

### ELIMINAR COLUMNAS
Creo que para este análisis podemos prescindir de la informacion sobre vuelos cancelados o desviados, el tiempo que pasa el avión en pista con los pasajeros dentro (TaxiIn, TaxiOut), la distancia y el destino del vuelo, el tiempo que ha pasado el avion en el aire, el número de vuelo y tambien la columna del año (todas son 2008) y unnamed.

In [9]:
columnas_para_eliminar = ['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 
                          'FlightNum', 'Cancelled', 'CancellationCode', 'Diverted', 'TaxiIn', 'TaxiOut']
airlines_delay_no_nan = airlines_delay_no_nan.drop(columnas_para_eliminar, axis=1)
airlines_delay_no_nan

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier TailNum  \
3         1829.0        1755   1959.0        1925            WN  N464WN   
5         1937.0        1830   2037.0        1940            WN  N763SW   
7         1644.0        1510   1845.0        1725            WN  N334SW   
9         1452.0        1425   1640.0        1625            WN  N286WN   
11        1323.0        1255   1526.0        1510            WN  N674AA   
...          ...         ...      ...         ...           ...     ...   
1936751    921.0         830   1112.0        1008            DL  N907DE   
1936752   1552.0        1520   1735.0        1718            DL  N905DE   
1936753   1250.0        1220   1617.0        1552            DL  N938DL   
1936754    657.0         600    904.0         749            DL  N3743H   
1936755   1007.0         847   1149.0        1010            DL  N909DA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
3                     90.0            90.0     77.0      34.0      34.0   
5                    240.0           250.0    230.0      57.0      67.0   
7                    121.0           135.0    107.0      80.0      94.0   
9                    228.0           240.0    213.0      15.0      27.0   
11                   123.0           135.0    110.0      16.0      28.0   
...                    ...             ...      ...       ...       ...   
1936751              111.0            98.0     82.0      64.0      51.0   
1936752               43.0            58.0     27.0      17.0      32.0   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
3          IND  BWI       515           2.0           0.0       0.0   
5          IND  LAS      1591          10.0           0.0       0.0   
7          IND  MCO       828           8.0           0.0       0.0   
9          IND  PHX      1489           3.0           0.0       0.0   
11         IND  TPA       838           0.0           0.0       0.0   
...        ...  ...       ...           ...           ...       ...   
1936751    ATL  PBI       545          51.0           0.0      13.0   
1936752    HSV  ATL       151           0.0           0.0       0.0   
1936753    MSP  ATL       906           3.0           0.0       0.0   
1936754    RIC  ATL       481           0.0          57.0      18.0   
1936755    ATL  IAH       689           1.0           0.0      19.0   

         SecurityDelay  LateAircraftDelay  
3                  0.0               32.0  
5                  0.0               47.0  
7                  0.0               72.0  
9                  0.0               12.0  
11                 0.0               16.0  
...                ...                ...  
1936751            0.0                0.0  
1936752            0.0               17.0  
1936753            0.0               22.0  
1936754            0.0                0.0  
1936755            0.0               79.0  

[1247486 rows x 19 columns]

### TIPO DE DATO
Primero compruebo qué tipos de dato contiene el dataset.

In [10]:
tipos_de_datos = airlines_delay_no_nan.dtypes
print(tipos_de_datos)

DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


In [11]:
def float_a_int(valor):
    if isinstance(valor, float):
        return int(valor)
    else:
        return valor

airlines_delay_no_nan = airlines_delay_no_nan.applymap(float_a_int)
airlines_delay_no_nan

DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier TailNum  \
3           1829        1755     1959        1925            WN  N464WN   
5           1937        1830     2037        1940            WN  N763SW   
7           1644        1510     1845        1725            WN  N334SW   
9           1452        1425     1640        1625            WN  N286WN   
11          1323        1255     1526        1510            WN  N674AA   
...          ...         ...      ...         ...           ...     ...   
1936751      921         830     1112        1008            DL  N907DE   
1936752     1552        1520     1735        1718            DL  N905DE   
1936753     1250        1220     1617        1552            DL  N938DL   
1936754      657         600      904         749            DL  N3743H   
1936755     1007         847     1149        1010            DL  N909DA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
3                       90              90       77        34        34   
5                      240             250      230        57        67   
7                      121             135      107        80        94   
9                      228             240      213        15        27   
11                     123             135      110        16        28   
...                    ...             ...      ...       ...       ...   
1936751                111              98       82        64        51   
1936752                 43              58       27        17        32   
1936753                147             152      120        25        30   
1936754                127             109       78        75        57   
1936755                162             143      122        99        80   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
3          IND  BWI       515             2             0         0   
5          IND  LAS      1591            10             0         0   
7          IND  MCO       828             8             0         0   
9          IND  PHX      1489             3             0         0   
11         IND  TPA       838             0             0         0   
...        ...  ...       ...           ...           ...       ...   
1936751    ATL  PBI       545            51             0        13   
1936752    HSV  ATL       151             0             0         0   
1936753    MSP  ATL       906             3             0         0   
1936754    RIC  ATL       481             0            57        18   
1936755    ATL  IAH       689             1             0        19   

         SecurityDelay  LateAircraftDelay  
3                    0                 32  
5                    0                 47  
7                    0                 72  
9                    0                 12  
11                   0                 16  
...                ...                ...  
1936751              0                  0  
1936752              0                 17  
1936753              0                 22  
1936754              0                  0  
1936755              0                 79  

[1247486 rows x 19 columns]

In [12]:
columnas_horarias = ['DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime']

for columna in columnas_horarias:
    airlines_delay_no_nan[columna] = pd.to_datetime(airlines_delay_no_nan[columna], format='%H%M', errors='coerce')
for columna in columnas_horarias:
    airlines_delay_no_nan[columna] = airlines_delay_no_nan[columna].dt.strftime('%H:%M')

# Ahora airlines_delay tiene nuevas columnas que contienen solo la hora
airlines_delay_no_nan


DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
3         18:29      17:55   19:59      19:25            WN  N464WN   
5         19:37      18:30   20:37      19:40            WN  N763SW   
7         16:44      15:10   18:45      17:25            WN  N334SW   
9         14:52      14:25   16:40      16:25            WN  N286WN   
11        13:23      12:55   15:26      15:10            WN  N674AA   
...         ...        ...     ...        ...           ...     ...   
1936751   09:21      08:30   11:12      10:08            DL  N907DE   
1936752   15:52      15:20   17:35      17:18            DL  N905DE   
1936753   12:50      12:20   16:17      15:52            DL  N938DL   
1936754   06:57      06:00   09:04      07:49            DL  N3743H   
1936755   10:07      08:47   11:49      10:10            DL  N909DA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
3                       90              90       77        34        34   
5                      240             250      230        57        67   
7                      121             135      107        80        94   
9                      228             240      213        15        27   
11                     123             135      110        16        28   
...                    ...             ...      ...       ...       ...   
1936751                111              98       82        64        51   
1936752                 43              58       27        17        32   
1936753                147             152      120        25        30   
1936754                127             109       78        75        57   
1936755                162             143      122        99        80   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
3          IND  BWI       515             2             0         0   
5          IND  LAS      1591            10             0         0   
7          IND  MCO       828             8             0         0   
9          IND  PHX      1489             3             0         0   
11         IND  TPA       838             0             0         0   
...        ...  ...       ...           ...           ...       ...   
1936751    ATL  PBI       545            51             0        13   
1936752    HSV  ATL       151             0             0         0   
1936753    MSP  ATL       906             3             0         0   
1936754    RIC  ATL       481             0            57        18   
1936755    ATL  IAH       689             1             0        19   

         SecurityDelay  LateAircraftDelay  
3                    0                 32  
5                    0                 47  
7                    0                 72  
9                    0                 12  
11                   0                 16  
...                ...                ...  
1936751              0                  0  
1936752              0                 17  
1936753              0                 22  
1936754              0                  0  
1936755              0                 79  

[1247486 rows x 19 columns]

In [13]:
valores_nulos_por_columnas = airlines_delay_no_nan.isnull().sum()
print (valores_nulos_por_columnas)

DepTime               2336
CRSDepTime              35
ArrTime              11089
CRSArrTime            3148
UniqueCarrier            0
TailNum                  0
ActualElapsedTime        0
CRSElapsedTime           0
AirTime                  0
ArrDelay                 0
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
CarrierDelay             0
WeatherDelay             0
NASDelay                 0
SecurityDelay            0
LateAircraftDelay        0
dtype: int64


Tras la conversión a formato datetime se vuelven a generar valores nulos.

In [15]:
airlines_delay_def=airlines_delay_no_nan.dropna()
airlines_delay_def

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
3         18:29      17:55   19:59      19:25            WN  N464WN   
5         19:37      18:30   20:37      19:40            WN  N763SW   
7         16:44      15:10   18:45      17:25            WN  N334SW   
9         14:52      14:25   16:40      16:25            WN  N286WN   
11        13:23      12:55   15:26      15:10            WN  N674AA   
...         ...        ...     ...        ...           ...     ...   
1936751   09:21      08:30   11:12      10:08            DL  N907DE   
1936752   15:52      15:20   17:35      17:18            DL  N905DE   
1936753   12:50      12:20   16:17      15:52            DL  N938DL   
1936754   06:57      06:00   09:04      07:49            DL  N3743H   
1936755   10:07      08:47   11:49      10:10            DL  N909DA   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
3                       90              90       77        34        34   
5                      240             250      230        57        67   
7                      121             135      107        80        94   
9                      228             240      213        15        27   
11                     123             135      110        16        28   
...                    ...             ...      ...       ...       ...   
1936751                111              98       82        64        51   
1936752                 43              58       27        17        32   
1936753                147             152      120        25        30   
1936754                127             109       78        75        57   
1936755                162             143      122        99        80   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
3          IND  BWI       515             2             0         0   
5          IND  LAS      1591            10             0         0   
7          IND  MCO       828             8             0         0   
9          IND  PHX      1489             3             0         0   
11         IND  TPA       838             0             0         0   
...        ...  ...       ...           ...           ...       ...   
1936751    ATL  PBI       545            51             0        13   
1936752    HSV  ATL       151             0             0         0   
1936753    MSP  ATL       906             3             0         0   
1936754    RIC  ATL       481             0            57        18   
1936755    ATL  IAH       689             1             0        19   

         SecurityDelay  LateAircraftDelay  
3                    0                 32  
5                    0                 47  
7                    0                 72  
9                    0                 12  
11                   0                 16  
...                ...                ...  
1936751              0                  0  
1936752              0                 17  
1936753              0                 22  
1936754              0                  0  
1936755              0                 79  

[1230957 rows x 19 columns]

### MUESTRA ALEATORIA

Usamos una muestra aleatoria de 200.000 registros. Para asegurar que cada vez que se ejecute el código sea igual, utilizo el metodo random_state para fijar la semilla a partir de la cual genera 'muestra_airlines_delay'. 

In [16]:
muestra_airlines_delay = airlines_delay_def.sample(n=200000, random_state=42)
muestra_airlines_delay

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
1602716   19:18      18:10   21:34      20:35            9E  86729E   
1928810   20:18      17:50   22:11      19:34            CO  N56859   
439110    18:59      18:35   20:40      20:15            OO  N308SW   
940891    09:00      08:45   12:20      11:58            OH  N641CA   
669029    10:38      10:00   12:04      11:35            MQ  N902BC   
...         ...        ...     ...        ...           ...     ...   
1705041   10:13      09:41   12:43      12:27            NW  N523US   
1640263   07:16      06:50   09:37      09:20            WN  N713SW   
783947    09:49      09:20   11:19      10:55            OO  N983CA   
1582549   13:00      12:15   14:30      13:18            EV  N633AS   
1500468   13:10      12:35   15:04      14:34            9E  85159E   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
1602716                136             145       89        59        68   
1928810                233             224      204       157       148   
439110                 101             100       81        25        24   
940891                 200             193      161        22        15   
669029                  86              95       70        29        38   
...                    ...             ...      ...       ...       ...   
1705041                150             166      128        16        32   
1640263                321             330      309        17        26   
783947                  90              95       73        24        29   
1582549                 90              63       43        72        45   
1500468                114             119       83        30        35   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
1602716    IND  LGA       659             0             0        59   
1928810    IAH  LAX      1379             0             6         9   
439110     CPR  SLC       320             0             0         0   
940891     LGA  RSW      1080             0             0        22   
669029     DFW  MOB       540             8             0         0   
...        ...  ...       ...           ...           ...       ...   
1705041    MCO  DTW       957            16             0         0   
1640263    BWI  LAS      2106            17             0         0   
783947     BNA  MKE       475             0             0         0   
1582549    ATL  AVL       164            45             0        27   
1500468    ALB  DTW       488             0             0         0   

         SecurityDelay  LateAircraftDelay  
1602716              0                  0  
1928810              0                142  
439110               0                 25  
940891               0                  0  
669029               0                 21  
...                ...                ...  
1705041              0                  0  
1640263              0                  0  
783947               0                 24  
1582549              0                  0  
1500468              0                 30  

[200000 rows x 19 columns]

# EXERCICI 2

## Fes un informe complet del dataset:

#### Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

In [17]:
resumen_estadistico=muestra_airlines_delay.describe()
resumen_estadistico

ActualElapsedTime  CRSElapsedTime        AirTime       ArrDelay  \
count       200000.00000   200000.000000  200000.000000  200000.000000   
mean           135.04842      131.427460     107.127865      63.082105   
std             71.98201       69.283113      67.883458      60.531052   
min             15.00000      -21.000000       0.000000      15.000000   
25%             82.00000       80.000000      58.000000      26.000000   
50%            118.00000      115.000000      89.000000      43.000000   
75%            167.00000      160.000000     136.000000      78.000000   
max            697.00000      660.000000     664.000000    1583.000000   

            DepDelay       Distance   CarrierDelay   WeatherDelay  \
count  200000.000000  200000.000000  200000.000000  200000.000000   
mean       59.461145     738.734335      19.116180       3.689285   
std        59.636432     556.848314      43.284714      21.458951   
min         6.000000      31.000000       0.000000       0.000000   
25%        24.000000     334.000000       0.000000       0.000000   
50%        41.000000     594.000000       2.000000       0.000000   
75%        75.000000     967.000000      21.000000       0.000000   
max      1552.000000    4962.000000    1552.000000    1094.000000   

            NASDelay  SecurityDelay  LateAircraftDelay  
count  200000.000000  200000.000000      200000.000000  
mean       15.081840       0.088350          25.106450  
std        34.005003       1.908479          41.932449  
min         0.000000       0.000000           0.000000  
25%         0.000000       0.000000           0.000000  
50%         2.000000       0.000000           8.000000  
75%        15.000000       0.000000          33.000000  
max      1337.000000     216.000000        1184.000000

La función .describe() muestra un primer análisis estadístico. Podemos destacar que la media de la duración de los vuelos es de 135 minutos incluyendo el tiempo en pista de los que de media está 107 minutos en el aire. Con esto vemos que el tiempo medio que pasa un avión en pista con los pasajeros sin moverse es de unos 28 minutos. La distancia media es de 742 millas. También vemos los percentiles y encuentro valores muy altos en max. He tomado como ejemplo ArrDelay pensando que quizá algún tiempo se habia quedado en el "formato hora" en lugar de en minutos, pero al realizar sort_values el siguiente resultado era 1592 con lo que no puede ser horas y minutos. También he examinado algunas de las filas con valores mas altos pero no he visto nada que me haya resultado extraño. No sé cuál es la razón de los valores tan altos en max. 

### Troba quantes dades faltants hi ha per columna.

In [18]:
valores_nulos_por_columnas = muestra_airlines_delay.isnull().sum()
print (valores_nulos_por_columnas)

DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64


No hay ningún dato faltante en el dataset.

### Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
Compruebo que la distancia está en millas y el tiempo en minutos. La columna muestra la velocidad media en Mph.


In [19]:
muestra_airlines_delay['Velocidad_Media_mph'] = muestra_airlines_delay['Distance'] / (muestra_airlines_delay['AirTime'] / 60)
muestra_airlines_delay

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
1602716   19:18      18:10   21:34      20:35            9E  86729E   
1928810   20:18      17:50   22:11      19:34            CO  N56859   
439110    18:59      18:35   20:40      20:15            OO  N308SW   
940891    09:00      08:45   12:20      11:58            OH  N641CA   
669029    10:38      10:00   12:04      11:35            MQ  N902BC   
...         ...        ...     ...        ...           ...     ...   
1705041   10:13      09:41   12:43      12:27            NW  N523US   
1640263   07:16      06:50   09:37      09:20            WN  N713SW   
783947    09:49      09:20   11:19      10:55            OO  N983CA   
1582549   13:00      12:15   14:30      13:18            EV  N633AS   
1500468   13:10      12:35   15:04      14:34            9E  85159E   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
1602716                136             145       89        59        68   
1928810                233             224      204       157       148   
439110                 101             100       81        25        24   
940891                 200             193      161        22        15   
669029                  86              95       70        29        38   
...                    ...             ...      ...       ...       ...   
1705041                150             166      128        16        32   
1640263                321             330      309        17        26   
783947                  90              95       73        24        29   
1582549                 90              63       43        72        45   
1500468                114             119       83        30        35   

        Origin Dest  Distance  CarrierDelay  WeatherDelay  NASDelay  \
1602716    IND  LGA       659             0             0        59   
1928810    IAH  LAX      1379             0             6         9   
439110     CPR  SLC       320             0             0         0   
940891     LGA  RSW      1080             0             0        22   
669029     DFW  MOB       540             8             0         0   
...        ...  ...       ...           ...           ...       ...   
1705041    MCO  DTW       957            16             0         0   
1640263    BWI  LAS      2106            17             0         0   
783947     BNA  MKE       475             0             0         0   
1582549    ATL  AVL       164            45             0        27   
1500468    ALB  DTW       488             0             0         0   

         SecurityDelay  LateAircraftDelay  Velocidad_Media_mph  
1602716              0                  0           444.269663  
1928810              0                142           405.588235  
439110               0                 25           237.037037  
940891               0                  0           402.484472  
669029               0                 21           462.857143  
...                ...                ...                  ...  
1705041              0                  0           448.593750  
1640263              0                  0           408.932039  
783947               0                 24           390.410959  
1582549              0                  0           228.837209  
1500468              0                 30           352.771084  

[200000 rows x 20 columns]

Se puede observar que en los vuelos cortos la velocidad media es más baja que en los trayectos más largos. 

Ahora creo otra columna para mostrar si el vuelo se ha retrasado o no.

In [20]:
muestra_airlines_delay['Delayed'] = muestra_airlines_delay['ArrTime']>muestra_airlines_delay['CRSArrTime']
muestra_airlines_delay

DepTime CRSDepTime ArrTime CRSArrTime UniqueCarrier TailNum  \
1602716   19:18      18:10   21:34      20:35            9E  86729E   
1928810   20:18      17:50   22:11      19:34            CO  N56859   
439110    18:59      18:35   20:40      20:15            OO  N308SW   
940891    09:00      08:45   12:20      11:58            OH  N641CA   
669029    10:38      10:00   12:04      11:35            MQ  N902BC   
...         ...        ...     ...        ...           ...     ...   
1705041   10:13      09:41   12:43      12:27            NW  N523US   
1640263   07:16      06:50   09:37      09:20            WN  N713SW   
783947    09:49      09:20   11:19      10:55            OO  N983CA   
1582549   13:00      12:15   14:30      13:18            EV  N633AS   
1500468   13:10      12:35   15:04      14:34            9E  85159E   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  ...  Origin  \
1602716                136             145       89        59  ...     IND   
1928810                233             224      204       157  ...     IAH   
439110                 101             100       81        25  ...     CPR   
940891                 200             193      161        22  ...     LGA   
669029                  86              95       70        29  ...     DFW   
...                    ...             ...      ...       ...  ...     ...   
1705041                150             166      128        16  ...     MCO   
1640263                321             330      309        17  ...     BWI   
783947                  90              95       73        24  ...     BNA   
1582549                 90              63       43        72  ...     ATL   
1500468                114             119       83        30  ...     ALB   

        Dest Distance  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1602716  LGA      659             0             0        59              0   
1928810  LAX     1379             0             6         9              0   
439110   SLC      320             0             0         0              0   
940891   RSW     1080             0             0        22              0   
669029   MOB      540             8             0         0              0   
...      ...      ...           ...           ...       ...            ...   
1705041  DTW      957            16             0         0              0   
1640263  LAS     2106            17             0         0              0   
783947   MKE      475             0             0         0              0   
1582549  AVL      164            45             0        27              0   
1500468  DTW      488             0             0         0              0   

         LateAircraftDelay  Velocidad_Media_mph  Delayed  
1602716                  0           444.269663     True  
1928810                142           405.588235     True  
439110                  25           237.037037     True  
940891                   0           402.484472     True  
669029                  21           462.857143     True  
...                    ...                  ...      ...  
1705041                  0           448.593750     True  
1640263                  0           408.932039     True  
783947                  24           390.410959     True  
1582549                  0           228.837209     True  
1500468                 30           352.771084     True  

[200000 rows x 21 columns]

In [21]:
print (muestra_airlines_delay['Delayed'].value_counts())

True     189921
False     10079
Name: Delayed, dtype: int64


En la muestra aleatoria de 200.000 registros encuentro 189921 vuelos retrasados. Este dato supone que el 94.9% de los vuelos han sufrido retrasos.

### Fes una taula de les aerolínies amb més endarreriments acumulats.
Conteo de vuelos por cada aerolinea. A continuación se muestra el número de retrasos de cada una de ellas y el porcentaje. En la primera tabla se ordenan por número de vuelos de mayor a menor, y en la segunda tabla se muestran ordenadas por el porcentaje de retrasos de menor a mayor.

In [22]:
num_aerolineas = muestra_airlines_delay['UniqueCarrier'].value_counts()
num_retrasos = muestra_airlines_delay[muestra_airlines_delay['Delayed'] > 0].groupby('UniqueCarrier')['Delayed'].count()
retrasos_aerolineas = pd.DataFrame({'VuelosTotales': num_aerolineas, 'Retrasos': num_retrasos})
retrasos_aerolineas['PorcentajeRetrasos'] = (retrasos_aerolineas['Retrasos'] / retrasos_aerolineas['VuelosTotales']) * 100
retrasos_aerolineas = retrasos_aerolineas.sort_values(by='VuelosTotales', ascending=False)

retrasos_aerolineas

VuelosTotales  Retrasos  PorcentajeRetrasos
WN          32790     31350           95.608417
AA          21123     19773           93.608862
MQ          15676     15316           97.703496
UA          15239     14238           93.431328
OO          14260     13775           96.598878
DL          11628     11059           95.106639
XE          11532     10975           95.169962
US           9577      9039           94.382374
CO           9425      8664           91.925729
EV           9241      8979           97.164809
NW           8811      8395           95.278629
YV           8067      7672           95.103508
FL           7568      7149           94.463531
OH           6399      6152           96.140022
B6           6014      5288           87.928168
9E           5645      5485           97.165633
AS           3748      3582           95.570971
F9           2529      2338           92.447608
HA            678       642           94.690265
AQ             50        50          100.000000

In [23]:
retrasos_aerolineas = retrasos_aerolineas.sort_values(by='PorcentajeRetrasos', ascending=True)
retrasos_aerolineas

VuelosTotales  Retrasos  PorcentajeRetrasos
B6           6014      5288           87.928168
CO           9425      8664           91.925729
F9           2529      2338           92.447608
UA          15239     14238           93.431328
AA          21123     19773           93.608862
US           9577      9039           94.382374
FL           7568      7149           94.463531
HA            678       642           94.690265
YV           8067      7672           95.103508
DL          11628     11059           95.106639
XE          11532     10975           95.169962
NW           8811      8395           95.278629
AS           3748      3582           95.570971
WN          32790     31350           95.608417
OH           6399      6152           96.140022
OO          14260     13775           96.598878
EV           9241      8979           97.164809
9E           5645      5485           97.165633
MQ          15676     15316           97.703496
AQ             50        50          100.000000

In [24]:
retrasos = retrasos_aerolineas['PorcentajeRetrasos'].describe()
retrasos

count     20.000000
mean      94.974441
std        2.486711
min       87.928168
25%       94.188996
50%       95.138300
75%       96.254736
max      100.000000
Name: PorcentajeRetrasos, dtype: float64

De media, el 94.97% de los vuelos mostrados en el dataset se han retrasado.

La compañía que más vuelos retrasados acumula es Southwest Airlines (WN). Es una compañía de bajo coste que realiza principalmente vuelos dentro de los Estados Unidos. Su porcentaje de retraso se encuentra un 0.64% por encima de la media. Es una de las compañías con más tráfico aéreo del mundo y la que más representación tiene en el dataset, 32790 vuelos, un 16.34% de los datos de la muestra.

Como compañía con menos retrasos sobre el total de sus vuelos, tenemos JetBlue Airways (B6). Es otra compañía de bajo coste que también cubre algunos trayectos internacionales. Su porcentaje de retraso es del 87.93% viéndose afectados 5288 vuelos sobre un total de 6014. Esta sería la compañía más recomendable para intentar evitar retrasos, aunque es importante destacar que a nivel del total de vuelos, se encuentra en la posición número 15 de un total de 20.

Teniendo en cuenta no solo el porcentaje de retrasos, sino también el número de vuelos, posiblemente la compañía más recomendable sería American Airlines (AA), ya que siendo la segunda compañía que más vuelos opera, es la quinta mejor en porcentaje de retrasos. Otra compañía en situación similar es United Airlines (UA). En este caso es la cuarta compañía en cuanto a volumen de vuelos y también la cuarta que menos retrasos acumula.

Si tuviera que coger un vuelo en Estados Unidos, la primera opción que probaría sería JetBlue Airways. Si debido a su poca disposición de vuelos no fuera posible, las siguientes opciones serían AA o UA.

También sería recomendable Continental Airlines (CO), ya que es la segunda con mejor porcentaje y un número considerable de vuelos. El problema es que se unió a United Airlines en 2011 y ya no opera como compañía en solitario.

Como caso excepcional encontramos la aerolínea 9air (AQ). Es una aerolínea China que principalmente opera en ese país y sus vuelos internacionales (según su página web) se limitan a Laos y Tailandia. Necesitaría tener más información sobre la aerolínea en cuestión y su actividad en 2008. Por otro lado, representa tan solo el 0.025% del total de los vuelos, así que entre la falta de información y el poco volumen de vuelo, no la considero como la aerolínea con más retrasos a pesar de que el 100% de sus 50 vuelos en 2008 se retrasaron.

### Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.

In [25]:
rutas = muestra_airlines_delay.sort_values(by='Distance', ascending=False)
rutas_mas_largas = rutas[['Origin', 'Dest', 'Distance']].drop_duplicates(subset='Distance')

print(rutas_mas_largas)

        Origin Dest  Distance
556373     EWR  HNL      4962
1263822    ATL  HNL      4502
66592      ORD  HNL      4243
1323569    KOA  ORD      4213
1818597    ORD  OGG      4184
...        ...  ...       ...
942589     MHT  BOS        45
1057736    JNU  GST        41
941510     SRQ  TPA        40
53840      SNA  LAX        36
548896     WRG  PSG        31

[1364 rows x 3 columns]


La ruta más larga es de Newark (EWR) a Honolulu (HNL) con 4962 millas.

In [26]:
retrasos_acumulados = muestra_airlines_delay.groupby(['Origin', 'Dest'])['Delayed'].sum().reset_index()
rutas_con_mas_retrasos = retrasos_acumulados.sort_values(by='Delayed', ascending=False)

print(rutas_con_mas_retrasos)

     Origin Dest  Delayed
2505    LAX  SFO      494
3455    ORD  LGA      488
4241    SFO  LAX      441
212     ATL  LGA      431
2577    LGA  ORD      383
...     ...  ...      ...
3177    MSP  PWM        0
2979    MIA  PIT        0
2372    LAS  EUG        0
3038    MKG  GRR        0
4377    SLC  FAI        0

[4788 rows x 3 columns]


Los trayectos que más retrasos acumulan son Los Angeles (LAX) San Francisco (SFO) con 494 retrasos, después viene el trayecto Chicago (ORD) New York (LAG) con 488 vuelos, en tercera posición vendría el mismo trayecto ubicado en primera posición, pero al revés (SFO)-(LAX) y el cuarto es el que cubre el trayecto Atlanta (ATL) New York (LGA) con 431 retrasos acumulados.

Son aeropuertos muy transitados. El de Atlanta (ATL) es el de más tráfico aéreo del mundo, el cuarto en ese ranking es el de Chicago (ORD), y Los Angeles (LAX) se encuentra en la sexta posición.

Con lo que de los cinco aeropuertos implicados en los cuatro trayectos que más retrasos acumulan en nuestra muestra, tres de ellos están entre los 10 más transitados del mundo. Creo que este puede ser uno de los motivos de que tengan más retrasos.

Como aportación extra, he dejado la columna TailNum para ver si hay alguna relación etre aviones y retrasos. Para ello hago un conteo de las veces que aparece cada avión como 'Delayed'.

In [38]:
retrasos_por_avion = muestra_airlines_delay.groupby('TailNum')['Delayed'].count().reset_index()
retrasos_por_avion = retrasos_por_avion.sort_values(by='Delayed', ascending=False)

print(retrasos_por_avion)

     TailNum  Delayed
1524  N37342      127
966   N27173      113
1501  N37208      110
603   N17175      109
2667  N570ML      105
...      ...      ...
820   N228UA        1
813   N226UA        1
245   N119UA        1
801   N221UA        1
5252  NHZOAL        1

[5253 rows x 2 columns]


Con esto sabemos que hay 5253 aviones distintos en el dataset, pero no se ver más información sobre si alguno de ellos tiene más tendencia a retrasarse o no. 

In [41]:
muestra_airlines_delay.to_excel('estructures_Dataframe.xlsx', index=False)